ts/src/pro/coinbase.ts

```diff
@@ -0,0 +1,560 @@
+'use strict';
+
+//  ---------------------------------------------------------------------------
+
+import coinbaseRest from '../coinbase.js';
+import { ArgumentsRequired } from '../base/errors.js';
+import { ArrayCacheBySymbolById } from '../base/ws/Cache.js';
+import { sha256 } from '../static_dependencies/noble-hashes/sha256.js';
+
+//  ---------------------------------------------------------------------------
+
+export default class coinbase extends coinbaseRest {
+    describe () {
+        return this.deepExtend (super.describe (), {
+            'has': {
+                'ws': true,
+                'watchOHLCV': false,
+                'watchOrderBook': true,
+                'watchTicker': true,
+                'watchTickers': true,
+                'watchTrades': true,
+                'watchBalance': false,
+                'watchOrders': true,
+                'watchMyTrades': false,
+            },
+            'urls': {
+                'api': {
+                    'ws': 'wss://advanced-trade-ws.coinbase.com',
+                },
+            },
+            'options': {
+                'tradesLimit': 1000,
+                'ordersLimit': 1000,
+                'myTradesLimit': 1000,
+                'sides': {
+                    'bid': 'bids',
+                    'offer': 'asks',
+                },
+            },
+        });
+    }
+
+    async subscribe (name, symbol = undefined, params = {}) {
+        /**
+         * @ignore
+         * @method
+         * @description subscribes to a websocket channel
+         * @see https://docs.cloud.coinbase.com/advanced-trade-api/docs/ws-overview#subscribe
+         * @param {string} name the name of the channel
+         * @param {string} symbol unified market symbol
+         * @param {object} params extra parameters specific to the cex api endpoint
+         * @returns {object} subscription to a websocket channel
+         */
+        await this.loadMarkets ();
+        let market = undefined;
+        let messageHash = name;
+        let productIds = [];
+        if (Array.isArray (symbol)) {
+            const symbols = this.marketSymbols (symbol);
```

Comment:
@samgermain same thing as we discussed before, if you don't make `symbols` part of the message hash, how would you properly resolve two different `watchTickers` calls?

```Javascript
const a = exchange.watchTickers(['BTC/USDT', 'LTC/USDT']) // messageHash = 'ticker_batch'
const b = exchange.watchTickers(['ADA/USDT', 'XRP/USDT']) // // messageHash = 'ticker_batch'
```

Reply:
right, sorry, I started this PR a while ago, I'll fix that

Reply:
@samgermain I think that is not a problem we have other exchanges like that, we just can't be resolving promises with symbols that do not belong to the subscription, but resolving one at a time it's ok